xarray_beam/_src/pangeo_forge.py

```diff
@@ -89,62 +107,38 @@ def __init__(
-    }
-    chunks = core.normalize_expanded_chunks(input_chunks, dim_sizes)
-    for key, (index, _) in zip(core.iter_chunk_keys(chunks),
-                               self.pattern.items()):
-      yield key, index
-
-  def _open_chunks(
-      self,
-      key: core.Key,
-      index: Tuple[int, ...]
-  ) -> Iterator[Tuple[core.Key, xarray.Dataset]]:
+    if pattern.merge_dims:
+      raise ValueError("patterns with `MergeDim`s are not supported.")
+
+  def _open_chunks(self, _) -> Iterator[Tuple[core.Key, xarray.Dataset]]:
     """Open datasets into chunks with XArray."""
-    path = self.pattern[index]
-    with FileSystems().open(path) as file:
-      dataset = xarray.open_dataset(
-        file, chunks=self.sub_chunks, **self.xarray_open_kwargs
-      )
-      dataset = _expand_dimensions_by_key(dataset, key, index, self.pattern)
+    for index, path in self.pattern.items():
+      with FileSystems().open(path) as file:
+        key = _pattern_index_to_key(index)
 
-      base_key = core.Key(_zero_dimensions(dataset)).with_offsets(**key.offsets)
+        dataset = xarray.open_dataset(
+          file, chunks=self.sub_chunks, **self.xarray_open_kwargs
+        )
+        dataset = _expand_dimensions_by_key(dataset, key, index, self.pattern)
 
-      num_threads = len(dataset.data_vars)
+        base_key = core.Key(_zero_dimensions(dataset)).with_offsets(
+          **key.offsets
+        )
 
-      if self.sub_chunks == -1:
-        yield base_key, dataset.compute(num_workers=num_threads)
-        return
+        num_threads = len(dataset.data_vars)
 
-      dim_sizes = {dim: dataset.dims[dim] for dim in self.sub_chunks.keys()}
-      norm_sub_chunks = core.normalize_expanded_chunks(
-        self.sub_chunks, cast(Mapping[str, int], dim_sizes)
-      )
-      offset_index = core.compute_offset_index(
-        core._chunks_to_offsets(norm_sub_chunks)
-      )
-      for sub_key in core.iter_chunk_keys(norm_sub_chunks):
-        sizes = {
-          dim: norm_sub_chunks[dim][offset_index[dim][offset]]
-          for dim, offset in sub_key.offsets.items()
-        }
-        slices = core.offsets_to_slices(
-          sub_key.offsets, sizes=sizes, base=base_key.offsets
-        )
-        chunk = dataset.isel(slices)
+        if self.sub_chunks == -1:
+          yield base_key, dataset.compute(num_workers=num_threads)
+          return
 
-        new_key = base_key.with_offsets(**sub_key.offsets)
-        yield new_key, chunk.chunk().compute(num_workers=num_threads)
+        for new_key, chunk in rechunk.split_chunks(base_key, dataset,
+                                                   self.sub_chunks):
+          yield new_key, chunk.compute(num_workers=num_threads)
 
   def expand(self, pcoll):
     return (
         pcoll
-        | beam.Create(self._prechunk())
-        | beam.FlatMapTuple(self._open_chunks)
+        | beam.Create([None])
+        | beam.FlatMap(self._open_chunks)
```

Comment:
I suggest using `beam.Create` directly:
```suggestion
        | beam.Create(self._open_chunks)
```

Why? `beam.Create` is "splittable", which means that if one of these opening tasks (or a fused downstream task) is slow, Beam can split up the work into smaller pieces. That isn't the case for FlatMap.

Reply:
TIL that `Create` can take a function instead of a collection! Thanks for the tip + explanation, I'm just learning about splittable do fns.

Reply:
Hmm... I've found when running tests that `Create` does indeed need to take an iterable (please check me on this). I remember now: I tried this before and it slowed down my local tests (which I notice once again). 

My best guess as to what is going on: The `_open_chunks` function is invoked at the time the `Create` node of the graph is created, instead of on each dataflow worker. Is there a better pattern I can use here?

Reply:
Here's an idea for a slight pivot: I follow the programming guide's [pattern for splittable do fns](https://beam.apache.org/documentation/programming-guide/#a-basic-sdf), where I define splits and restrictions that work with the sub_chunk / rechunk flow in `_open_chunks`. This will require re-working of this PTranform. 

Alternatively, I could save implementing this for a (near) future time.